ocf_data_sampler/torch_datasets/site.py

```diff
@@ -0,0 +1,271 @@
+
+            if nwp_config.max_staleness_minutes is None:
+                max_staleness = None
+            else:
+                max_staleness = minutes(nwp_config.max_staleness_minutes)
+
+            # The last step of the forecast is lost if we have to diff channels
+            if len(nwp_config.nwp_accum_channels) > 0:
+                end_buffer = pd.to_timedelta(nwp_config.time_resolution_minutes)
+            else:
+                end_buffer =minutes(0)
+
+            # This is the max staleness we can use considering the max step of the input data
+            max_possible_staleness = (
+                    pd.Timedelta(da["step"].max().item())
+                    - minutes(nwp_config.forecast_minutes)
+                    - end_buffer
+            )
+
+            # Default to use max possible staleness unless specified in config
+            if max_staleness is None:
+                max_staleness = max_possible_staleness
+            else:
+                # Make sure the max acceptable staleness isn't longer than the max possible
+                assert max_staleness <= max_possible_staleness
+
+            time_periods = find_contiguous_t0_periods_nwp(
+                datetimes=pd.DatetimeIndex(da["init_time_utc"]),
+                history_duration=minutes(nwp_config.history_minutes),
+                max_staleness=max_staleness,
+                max_dropout=max_dropout,
+            )
+
+            contiguous_time_periods[f'nwp_{nwp_key}'] = time_periods
+
+    if "sat" in datasets_dict:
+        sat_config = config.input_data.satellite
+
+        time_periods = find_contiguous_t0_periods(
+            pd.DatetimeIndex(datasets_dict["sat"]["time_utc"]),
+            sample_period_duration=minutes(sat_config.time_resolution_minutes),
+            history_duration=minutes(sat_config.history_minutes),
+            forecast_duration=minutes(sat_config.forecast_minutes),
+        )
+
+        contiguous_time_periods['sat'] = time_periods
+
+    # just get the values (not the keys)
+    contiguous_time_periods_values = list(contiguous_time_periods.values())
+
+    # Find joint overlapping contiguous time periods
+    if len(contiguous_time_periods_values) > 1:
+        valid_time_periods = intersection_of_multiple_dataframes_of_periods(
+            contiguous_time_periods_values
+        )
+    else:
+        valid_time_periods = contiguous_time_periods_values[0]
+
+    # check there are some valid time periods
+    if len(valid_time_periods) == 0:
+        raise ValueError(f"No valid time periods found, {contiguous_time_periods=}")
+
+    # 4. Now lets loop over each location in system id and find the valid periods
+    # Should we have a different option if there are not nans
+    sites = datasets_dict["site"]
+    site_ids = sites.site_id.values
+    site_config = config.input_data.site
+    valid_t0_and_site_ids = []
+    for site_id in site_ids:
+        site = sites.sel(site_id=site_id)
```

Comment:
I think this part all the way till the end of the function can be its own function. Or we can even keep it under the same name and just remove all above and replace with `valid_time_periods = find_valid_t0_times(..)` where `find_valid_t0_times` is the function currently in pvnet_uk_regional which we can get out of it as a common function bc I think the code above is exactly the same. I think the intersection of t0s of all supplementary data is the only thing the code below uses which is  `valid_time_periods`

Reply:
Hmm, good point, there is deinfately some duplicate code. let me see if I can refactor it

Reply:
Ive refactored to clean this up